File: app/src/main/java/com/example/newsapp/data/database/DatabaseDataManager.java
package com.example.newsapp.data.database;

import android.content.Context;
import android.os.AsyncTask;
import android.util.Log;

import com.example.newsapp.R;

import com.example.newsapp.data.database.accessobjects.ArticleAccessObject;
import com.example.newsapp.data.database.accessobjects.SourcesAccessObject;
import com.example.newsapp.data.datamodels.DataResponse;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class DatabaseDataManager extends AsyncTask<Void, Void,Void> {

    private  ArticleAccessObject articleAccessObject;
    private  SourcesAccessObject sourcesAccessObject;

    private ApiManager apiManager;

    private  Context context;
    private String query;

    public DatabaseDataManager(Context context, AppDatabase appDatabase, String query) {
        articleAccessObject = appDatabase.articleAccessObject();
        sourcesAccessObject = appDatabase.sourcesAccessObject();
        this.context= context;
        this.query = query;
    }

    public DatabaseDataManager(Context context, AppDatabase appDatabase){
        this.context = context;
        sourcesAccessObject = appDatabase.sourcesAccessObject();
    }

    @Override
    protected Void doInBackground(final Void... params) {
        if (query == null){
            apiManager = new ApiManager(context);

            Call<DataResponse> sourceCall = apiManager.getSources();
            sourceCall.enqueue(new Callback<DataResponse>() {
                @Override
                public void onResponse(Call<DataResponse> call, Response<DataResponse> sourceResponse) {

                    if (sourceResponse.body() != null){
                        for (int index = 0; index < sourceResponse.body().getSources().size() ; index++) {
                            sourcesAccessObject.createSourceDataIfNotExists(sourceResponse.body()
                                    .getSources().get(index));
                        }
                    }
                }

                @Override
                public void onFailure(Call<DataResponse> call, Throwable t) {
                    Log.d(context.getString(R.string.database_manager_sources_error),
                            t.getLocalizedMessage());
                }
            });



        } else {

            Log.d("datbase manafer",query);

            apiManager = new ApiManager(context, query);
            Call<DataResponse> call = apiManager.getArticles();
            call.enqueue(new Callback<DataResponse>() {
                @Override
                public void onResponse(Call<DataResponse> call, Response<DataResponse> response) {


                    if (response.body() != null) {

                        /* Consider removing loop*/

                        Log.d("Data gotten:", String.valueOf(response.body()));

                        for (int i = 0; i < response.body().getArticles().size(); i++) {

                            articleAccessObject.createDataIfNotExists(response.body().getArticles().get(i));
                        }
                    }
                }

                @Override
                public void onFailure(Call<DataResponse> call, Throwable t) {
                    Log.d("Data", t.getLocalizedMessage());

                }
            });
        }
        return null;
    }
}

File: app/src/main/java/com/example/newsapp/data/database/DatabaseDataManager.kt
package com.example.newsapp.data.database

import android.annotation.SuppressLint
import android.content.Context
import android.os.AsyncTask
import android.util.Log
import com.example.newsapp.BuildConfig
import com.example.newsapp.R
import com.example.newsapp.data.api.NetworkService
import com.example.newsapp.data.database.accessobjects.ArticleAccessObject
import com.example.newsapp.data.database.accessobjects.SourcesAccessObject
import com.example.newsapp.data.datamodels.DataResponse
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import javax.inject.Inject

class DatabaseDataManager : AsyncTask<Void, Void, Void> {

    private lateinit var articleAccessObject: ArticleAccessObject
    private var sourcesAccessObject: SourcesAccessObject? = null

    @Inject
    lateinit var apiManager: NetworkService

    @SuppressLint("StaticFieldLeak")
    @Inject
    lateinit var context: Context

    private lateinit var query: String

    constructor(context: Context, appDatabase: AppDatabase, query: String) {
        articleAccessObject = appDatabase.articleAccessObject()
        sourcesAccessObject = appDatabase.sourcesAccessObject()
        this.query = query
    }

    constructor(context: Context, appDatabase: AppDatabase) {
        this.context = context
        sourcesAccessObject = appDatabase.sourcesAccessObject()
    }

    override fun doInBackground(vararg params: Void): Void? {
        if (query == null) {

            val sourceCall = apiManager.getSources(BuildConfig.ApiKey)
            sourceCall.enqueue(object : Callback<DataResponse> {
                override fun onResponse(call: Call<DataResponse>, sourceResponse: Response<DataResponse>) {

                    if (sourceResponse.body() != null) {
                        for (index in 0 until sourceResponse.body()!!.sources!!.size) {
                            sourcesAccessObject!!.createSourceDataIfNotExists(sourceResponse.body()!!
                                    .sources!![index])
                        }
                    }
                }

                override fun onFailure(call: Call<DataResponse>, t: Throwable) {
                    Log.d(context.getString(R.string.database_manager_sources_error),
                            t.localizedMessage)
                }
            })


        } else {
            val call = apiManager.getAllNews(query, BuildConfig.ApiKey)
            call.enqueue(object : Callback<DataResponse> {
                override fun onResponse(call: Call<DataResponse>, response: Response<DataResponse>) {


                    if (response.body() != null) {

                        for (i in 0 until response.body()!!.articles!!.size) {

                            articleAccessObject.createDataIfNotExists(response.body()!!.articles!![i])
                        }
                    }
                }

                override fun onFailure(call: Call<DataResponse>, t: Throwable) {
                    Log.d("Data", t.localizedMessage)

                }
            })
        }
        return null
    }
}
